admin/server/auth/claims.go

```diff
@@ -247,3 +190,112 @@ func unionProjectRoles(a *adminv1.ProjectPermissions, b *database.ProjectRole) *
 		ManageProjectMembers: a.ManageProjectMembers || b.ManageProjectMembers,
 	}
 }
+
+// organizationPermissionsUser is a Claims implementation for resolve the organization permissions for user
+func (c *authTokenClaims) organizationPermissionsUser(ctx context.Context, orgID string) *adminv1.OrganizationPermissions {
+	c.Lock()
+	defer c.Unlock()
+
+	if perm, ok := c.orgPermissionsCache[orgID]; ok {
+		return perm
+	}
+
+	roles, err := c.admin.DB.ResolveOrganizationRolesForUser(context.Background(), c.token.OwnerID(), orgID)
+	if err != nil {
+		panic(fmt.Errorf("failed to get org permissions: %w", err))
+	}
+
+	composite := &adminv1.OrganizationPermissions{}
+	for _, role := range roles {
+		composite = unionOrgRoles(composite, role)
+	}
+
+	c.orgPermissionsCache[orgID] = composite
+	return composite
+}
+
+// organizationPermissionsService is a Claims implementation for resolve the organization permissions for service
+func (c *authTokenClaims) organizationPermissionsService(ctx context.Context, orgID string) *adminv1.OrganizationPermissions {
+	// Not adding caching for now
+	var orgPermissions *adminv1.OrganizationPermissions
+	service, err := c.admin.DB.FindService(ctx, c.token.OwnerID())
+	if err != nil {
+		panic(fmt.Errorf("failed to get service info: %w", err))
+	}
+
+	if orgID == service.OrgID {
+		orgPermissions = &adminv1.OrganizationPermissions{
+			ReadOrg:          true,
+			ManageOrg:        true,
+			ReadProjects:     true,
+			CreateProjects:   true,
+			ManageProjects:   true,
+			ReadOrgMembers:   true,
+			ManageOrgMembers: true,
+		}
+	}
+	return orgPermissions
+}
+
+// projectPermissionsUser is a Claims implementation for resolve the project permissions for user
+func (c *authTokenClaims) projectPermissionsUser(ctx context.Context, orgID, projectID string) *adminv1.ProjectPermissions {
+	// ManageProjects permission on the org gives full access to all projects in the org (only org admins have this)
+	orgPerms := c.OrganizationPermissions(ctx, orgID)
+	if orgPerms.ManageProjects {
+		return &adminv1.ProjectPermissions{
+			ReadProject:          true,
+			ManageProject:        true,
+			ReadProd:             true,
+			ReadProdStatus:       true,
+			ManageProd:           true,
+			ReadDev:              true,
+			ReadDevStatus:        true,
+			ManageDev:            true,
+			ReadProjectMembers:   true,
+			ManageProjectMembers: true,
+		}
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	if perm, ok := c.projectPermissionsCache[projectID]; ok {
+		return perm
+	}
```

Comment:
Same note, can be moved to common function so it works for services and users tokens